src/main/java/com/google/opensesame/projects/ProjectQuery.java

```diff
@@ -0,0 +1,127 @@
+    String[] filterRequests = request.getParameterValues(FILTER_QUERY_PARAM);
+    if (filterRequests == null) {
+      return Arrays.asList();
+    }
+
+    List<QueryFilter> queryFilters = new ArrayList<QueryFilter>();
+    for (String filterRequest : filterRequests) {
+      QueryFilter queryFilter = QueryFilter.fromString(filterRequest, response);
+      if (queryFilter == null) {
+        return null;
+      }
+      queryFilters.add(queryFilter);
+    }
+
+    return queryFilters;
+  }
+
+  /**
+   * Queries for ProjectEntities in the datastore based on a servlet request.
+   *
+   * @param request The servlet request defining the query.
+   * @param response The servlet response to send errors to.
+   * @return Returns a collection of ProjectEntities from the query or null if the query was
+   *     invalid.
+   * @throws IOException
+   */
+  public static Collection<ProjectEntity> queryFromRequest(
+      HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String[] projectIds = request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM);
+    Collection<ProjectEntity> projectEntities;
+    if (projectIds != null) {
+      projectEntities = ofy().load().type(ProjectEntity.class).ids(projectIds).values();
+
+      if (projectIds.length != projectEntities.size()) {
+        ErrorResponse.sendJsonError(
+            response,
+            "One or more of the supplied IDs is invalid and those projects could not be found.",
+            HttpServletResponse.SC_NOT_FOUND,
+            "One or more of the requested projects does not exist. Please try again later.");
+        return null;
+      }
+    } else {
+      // TODO(Richie): Add pagination support.
+      // TODO(Richie): Add ordering support.
+      Query<ProjectEntity> projectEntityQuery =
+          ofy().load().type(ProjectEntity.class).order("-numMentors");
+
+      List<QueryFilter> queryFilters = getQueryFiltersFromRequest(request, response);
+      if (queryFilters == null) {
+        return null;
+      }
+
+      for (QueryFilter queryFilter : queryFilters) {
+        projectEntityQuery =
+            projectEntityQuery.filter(queryFilter.condition, queryFilter.comparisonObject);
+      }
+
+      projectEntities = projectEntityQuery.list();
+    }
+
+    return projectEntities;
+  }
+
+  /**
+   * Get a project entity with a specified repository ID from the datastore or create a new one if
+   * one does not exist in the datastore.
+   *
+   * @param repositoryId The repository ID of the project.
+   * @return Returns the project entity from datastore or a new project entity.
+   */
+  public static ProjectEntity fromRepositoryIdOrNew(String repositoryId) {
+    ProjectEntity projectEntity = ofy().load().type(ProjectEntity.class).id(repositoryId).now();
+    if (projectEntity == null) {
+      projectEntity =
+          new ProjectEntity(repositoryId, new ArrayList<String>(), new ArrayList<String>());
+    }
+
+    return projectEntity;
+  }
```

Comment:
Is there a reason to move this function? Seems like it belongs in the ProjectEntity class to me.

Reply:
I think my reasoning was that it is technically "querying" the projects datastore, but on second thought I agree that it makes more sense in the other class. I'll move this class and its tests.